### PR TITLE
Use istio-iptables golang implementation instead of bash script

### PR DESCRIPTION
**Pull request by abhide**
Testing all E2E tests that will use golang implementation for istio-iptables
Issue: https://github.com/istio/istio/issues/13967

### CODE REVIEW

**Comment by rlenglet** [comment]
A bunch of tests don't like the additional option...

**Comment by abhide** [comment]
> A bunch of tests don't like the additional option...

Yeah.. Will take a look

**Comment by abhide** [comment]
Most (almost all) tests failures are related to pods unable to start due to images pulll

`image: "istio-testing/pilot:istio-testing-distroless"` 
`image: "istio-testing/mixer:istio-testing-distroless"`
`image: "istio-testing/galley:istio-testing-distroless"`

Looks like the distroless images aren't present or being looked up at wrong location?

@rlenglet @howardjohn Any ideas how to resolve this?

File: prow/e2e-kind-suite.sh [hunk 1]
```diff
@@ -105,6 +105,8 @@ if [[ "${ENABLE_ISTIO_CNI:-false}" == true ]]; then
    cni_run_daemon_kind
 fi
 
+
+VARIANT="distroless"
```
Comment (howardjohn) [inline prow/e2e-kind-suite.sh]: you need to move this before `build_images`

**Comment by abhide** [comment]
Thanks @howardjohn test results look better than before! Getting there!

**Comment by rlenglet** [comment]
This is the test that is failing:
https://github.com/istio/istio/blob/13cb3422ae47f95dfbc829a327724a56929cab3b/tests/e2e/tests/pilot/tcp_test.go#L106
And this is the particular command that is failing:
https://github.com/istio/istio/blob/13cb3422ae47f95dfbc829a327724a56929cab3b/tests/e2e/tests/pilot/tcp_test.go#L139-L140

The `app` container fails to start for some reason.

`kubelet` logged this about the `statefulset-0` pod:
```
Nov 25 05:27:39 istio-testing-control-plane kubelet[283]: E1125 05:27:39.314079     283 pod_workers.go:191] Error syncing pod 194a2fd2-249e-4261-be27-6a3aab126a85 ("statefulset-0_istio-system(194a2fd2-249e-4261-be27-6a3aab126a85)"), skipping: failed to "StartContainer" for "app" with ErrImagePull: "rpc error: code = Unknown desc = failed to pull and unpack image \"docker.io/istio-testing/app:istio-testing-distroless\": failed to resolve reference \"docker.io/istio-testing/app:istio-testing-distroless\": pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed"
Nov 25 05:27:39 istio-testing-control-plane containerd[83]: time="2019-11-25T05:27:39.478814816Z" level=info msg="ExecSync for \"fdc0b1118c3d3ed55e5190fde59012dd526a7ab19948fa917c26ed2c76704ab3\" with command [/usr/local/bin/galley probe --probe-path=/tmp/healthready --interval=10s] and timeout 1 (s)"
Nov 25 05:27:39 istio-testing-control-plane kubelet[283]: E1125 05:27:39.486508     283 pod_workers.go:191] Error syncing pod 194a2fd2-249e-4261-be27-6a3aab126a85 ("statefulset-0_istio-system(194a2fd2-249e-4261-be27-6a3aab126a85)"), skipping: failed to "StartContainer" for "app" with ImagePullBackOff: "Back-off pulling image \"istio-testing/app:istio-testing-distroless\""
...
Nov 25 05:27:40 istio-testing-control-plane kubelet[283]: E1125 05:27:40.492050     283 pod_workers.go:191] Error syncing pod 194a2fd2-249e-4261-be27-6a3aab126a85 ("statefulset-0_istio-system(194a2fd2-249e-4261-be27-6a3aab126a85)"), skipping: failed to "StartContainer" for "app" with ImagePullBackOff: "Back-off pulling image \"istio-testing/app:istio-testing-distroless\""
...
Nov 25 05:27:55 istio-testing-control-plane kubelet[283]: E1125 05:27:55.985478     283 pod_workers.go:191] Error syncing pod 194a2fd2-249e-4261-be27-6a3aab126a85 ("statefulset-0_istio-system(194a2fd2-249e-4261-be27-6a3aab126a85)"), skipping: failed to "StartContainer" for "app" with ErrImagePull: "rpc error: code = Unknown desc = failed to pull and unpack image \"docker.io/istio-testing/app:istio-testing-distroless\": failed to resolve reference \"docker.io/istio-testing/app:istio-testing-distroless\": pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed"
...
Nov 25 05:28:07 istio-testing-control-plane kubelet[283]: E1125 05:28:07.358790     283 pod_workers.go:191] Error syncing pod 194a2fd2-249e-4261-be27-6a3aab126a85 ("statefulset-0_istio-system(194a2fd2-249e-4261-be27-6a3aab126a85)"), skipping: failed to "StartContainer" for "app" with ImagePullBackOff: "Back-off pulling image \"istio-testing/app:istio-testing-distroless\""
```
etc. etc.

The `istio-testing/app:istio-testing-distroless` tag is obviously incorrect. `-distroless` shouldn't be appended to the tag. This is a test bug.

**Comment by rlenglet** [comment]
This bug is in https://github.com/istio/istio/blob/a408dbc296f7ba7a7b049325be2f2bea626fbfa1/tests/e2e/tests/pilot/pilot_test.go#L411

I believe this should call `tc.Kube.AppTag()` instead of `tc.Kube.PilotTag()`, cf. https://github.com/istio/istio/blob/a408dbc296f7ba7a7b049325be2f2bea626fbfa1/tests/e2e/tests/pilot/pilot_test.go#L382 (which didn't fail).

@abhide could you try this fix, and submit a proper PR if this works?

File: tools/istio-docker.mk [hunk 2]
```diff
@@ -60,11 +60,7 @@ $(ISTIO_DOCKER)/certs:
 
 # tell make which files are copied from the source tree and generate rules to copy them to the proper location:
 # TODO(sdake)                      $(NODE_AGENT_TEST_FILES) $(GRAFANA_FILES)
-DOCKER_FILES_FROM_SOURCE:=tools/packaging/common/istio-iptables.sh tools/packaging/common/istio-clean-iptables.sh \
-                          tests/testdata/certs/cert.crt tests/testdata/certs/cert.key tests/testdata/certs/cacert.pem
-# generates rules like the following:
```
Comment (rlenglet) [inline tools/istio-docker.mk]: Keep the comments. They explain the rules generated by the `foreach` below.

**Comment by abhide** [comment]
> This bug is in
> 
> https://github.com/istio/istio/blob/a408dbc296f7ba7a7b049325be2f2bea626fbfa1/tests/e2e/tests/pilot/pilot_test.go#L411
> 
> I believe this should call `tc.Kube.AppTag()` instead of `tc.Kube.PilotTag()`, cf.
> 
> https://github.com/istio/istio/blob/a408dbc296f7ba7a7b049325be2f2bea626fbfa1/tests/e2e/tests/pilot/pilot_test.go#L382
> 
> (which didn't fail).
> @abhide could you try this fix, and submit a proper PR if this works?

This is fixed in master branch by: https://github.com/istio/istio/pull/19252

**Comment by howardjohn** [comment]
actually I guess part of the motivation is that we don't have to maintain two implementation so maybe that's not worth it

**Comment by abhide** [comment]
> if we merge this should we copy the bash script in as well for a release as a backup in case there is some issue with the go impl? that way user could just switch the template to reference the shell script instead of waiting for a new release to fix issue

hmm.. don't we want to move away from bash script? We should fix any issues with golang implementation and get rid of bash script completely.

**Comment by rlenglet** [comment]
> actually I guess part of the motivation is that we don't have to maintain two implementation so maybe that's not worth it

Exactly what I'm thinking. The Golang implementation is now well covered by unit tests, so this is the implementation we should focus on.

File: tools/istio-docker.mk (hunk 2 above)
Comment (abhide) [reply tools/istio-docker.mk]: I think we don't need. Its already covered here: https://github.com/istio/istio/pull/18962/files#diff-d4ccc43f318f9d501d556d3a89f8f54dL52

**Comment by mandarjog** [comment]
Does golang impl already produce identical rules over the input space? 
We should totally get rid of bash and direct up tables stuff. How does distroless or non-distroless affect this?

**Comment by abhide** [comment]
> Does golang impl already produce identical rules over the input space?
> We should totally get rid of bash and direct up tables stuff. How does distroless or non-distroless affect this?

Golang uses iptables-restore instead of individual iptables commands. I tested out distroless there were 2 test issues

**Comment by howardjohn** [comment]
Mandar for some more info, distroless has already been using to go impl forever. This PR originally was switching to distroless default just to test the iptables stuff but now has switched to just using the go iptables everywhere

**Comment by howardjohn** [comment]
This lgtm, but will give others a chance to review

**Comment by kyessenov** [comment]
This is wonderful. We can slim the base image to drop C tools and statically link the go binary instead.

**Comment by howardjohn** [comment]
@kyessenov I think we still depend on iptables? See the distroless implementation, we copy these over

https://github.com/istio/istio/blob/718f156cf7919116cab035878ec6b51bd5e3f60b/pilot/docker/Dockerfile.proxyv2#L23-L26

But if we can clean some stuff up, would be great

**Comment by kyessenov** [comment]
OK, I'm confused. Is the golang implementation just a script wrapper to call out to iptables then?

**Comment by abhide** [comment]
> @kyessenov I think we still depend on iptables? See the distroless implementation, we copy these over
> 
> https://github.com/istio/istio/blob/718f156cf7919116cab035878ec6b51bd5e3f60b/pilot/docker/Dockerfile.proxyv2#L23-L26
> 
> But if we can clean some stuff up, would be great

We will need iptables, iptables-restore, iptables-save (and same for v6)
@kyessenov yes, golang implementation will invoke iptables-restore by default, but can also invoke individual iptables commands.

**Comment by kyessenov** [comment]
Do we have a plan to address the hack of copying binaries from one image to another? It's not safe to do so since we have varying versions of glibc used to build them.

**Comment by rlenglet** [comment]
@abhide please add an entry in the [Release 1.5 note draft](https://github.com/istio/istio/wiki/Istio-Release-1.5).

**Comment by abhide** [comment]
> @abhide please add an entry in the [Release 1.5 note draft](https://github.com/istio/istio/wiki/Istio-Release-1.5).

Added it under `**Other**` section

**Comment by ccccbjcn** [comment]
Do you have plan to support IPVS?

**Comment by rlenglet** [comment]
@ccccbjcn IPVS doesn't solve the same problem.

**Comment by ccccbjcn** [comment]
@rlenglet Thanks for your reply.  Yes, using IPVS still needs to write code to install the iptables. I mean whether using IPVS to improve performance for large scale APP deployment on a node?

**Pull request merged**